### ISSUE TITLE
Services may not be terminated in initrd breaking dependencies in real root

### ISSUE THREAD

**Issue by arvidjaar** [systemd/systemd#3436]
### Submission type
- [X] Bug report
- [ ] Request for enhancement (RFE)

_NOTE: Do not submit anything other than bug reports or RFEs via the issue tracker!_
### systemd version the issue has been seen with

> Patches 210 but I do not see any relevant changes in current upstream.

_NOTE: Do not submit bug reports about anything but the two most recently released systemd versions upstream!_
### Used distribution

> openSUSE 13.2
### In case of bug report: Expected behaviour you didn't see

> Before switching root in initrd all basic units (in particular basic.target and sysinit.target) are stopped and later started again in real root
### In case of bug report: Unexpected behaviour you saw

> Service that had been started in initrd are left in starting state when entering real root. As consequence, there is no synchronization between early boot services (before sysinit.target) and later units because most services wait only on special targets. Because special targets are already "started", both early and late boot services are started concurrently with unpredictable results.
### In case of bug report: Steps to reproduce the problem

> It is rather hard to trigger on purpose; so far it appeared on one user system and was provoked by accidental omission of initrd-udevadm-cleanup-db.service in initrd. I was not able to reproduce it on my VMs using the same version. See https://bugzilla.suse.com/show_bug.cgi?id=980324 for details of this problem.

Quoting above bug report:

Normally it is expected that units started in initrd are (mostly) stopped before switching to real root. But there is no real synchronization that guarantees it; it seems that the only service that delayed switch was `initrd-udevadm-cleanup-db.service`. Without this service (boo#980325) root switch happens too fast, recording basic.target/sysinit.target in Started state. This in turn means that everything that would normally wait for sysinit.target/basic.target is started concurrently with low level units, causing havoc.

Adding `After=initrd.target` to `initrd-switch-root.target` seems to fix it, but my concern is, this "fix" relies on very specific and known in advance services in initrd. Any third-party unit added to initrd nay trigger similar problem. We are facing again the question "how to write a unit that is guaranteed to run after all other units are stopped/started".

**Comment by poettering** [comment]
I am not sure I grok what you are trying to say. systemd will serialize the state across the initrd transition. A service that is in "starting" state at the time of the switch root will thus be in "starting" state after, too, and continue from there. If you want it to have finished starting before the transition, then add Before=initrd-switch-root.target to your unit.

note that initrd-cleanup.service uses "systemctl isolate initrd-switch-root.target" to initiate the transition, hence all services that shall remain around across the transition need to be a dependency of that target.

Anyway, really not grokking what your specific issue is?

**Comment by arvidjaar** [comment]
> really not grokking what your specific issue is?

Some units in inird shadow units in real system. Those units must be started twice - first time in initrd, second time in real system. Example of such units that we actually had problems with are sysinit.target, basic.target, kmod-static-node.service. Currently this is solved by (implicitly) stopping all units started in initrd when `systemctl isolate initrd-switch-root.target` is executed. The problem is, systemd performs stopping and starting in parallel unless _explicit_ dependencies are defined. Because most of units do not have explicit dependency on `initrd-switch-root.service`, it is executed in parallel with stopping other services. So it may happen that when root is switched, other services are still running. This means in real root systemd will skip starting these units (they will be missing in initial transaction). The effect is
- when basic.target is not part of initial transaction, there is no barrier between early and late boot services; they all are started concurrently. You may want to look at bug report I linked for results
- when kmod-static-nodes.service is not started second time in real root, only device nodes for modules present in initrd are created. Becase initrd of course includes only limited set of kernel modules, many of them will be missing.

> A service that is in "starting" state at the time of the switch root will thus be in "starting" state after, too

Where have you got "starting" state from? The problem is that those units are in _started_ state so are skipped in initial transaction.

> If you want it to have finished starting before the transition, then add Before=initrd-switch-root.target to your unit.

There is no "my" unit. The problem happens using only default units shipped by systemd. If this is your final answer to this problem, then please 
- fix systemd by adding Before to every unit used in initrd
- document this requirement that anyone creating units for use in initrd must add this Before line

But I yet another time will say that this does not scale and we need solution that allows us to express "this unit must be started only after everything else is stopped".

**Comment by esmil** [comment]
> note that initrd-cleanup.service uses "systemctl isolate initrd-switch-root.target" to initiate the transition, > hence all services that shall remain around across the transition need to be a dependency of that target.

This sounds good, but then there is a bug (probably race) in either the shipped units or the isolate logic of at least v227 and v231. I sometimes see local-fs.target (in the active state) cross the transition and as far as I can tell local-fs.target is not a dependency of initrd-switch-root.target.

**Comment by poettering** [comment]
Here's an idea. Maybe we should redefine what DefaultDependencies=yes means in the initrd. Right now pretty much all services running in the initrd set DefaultDependencies=no (or at least: should). We could hence permit DefaultDependencies=yes, and say that if we run in the initrd context it results in Conflicts=initrd-switch-root.service and Before=initrd-switch-root.service or something like that, the same way as DefaultDependencies=yes on the main host means, the same against shutdown.target. We could then replace the "systemctl isolate" invocation with a simple "systemctl start", and everything would be cleaned up automatically, unless it sets DefaultDependencies=no.

This should solve the issue you raise and make a lot of things easier, no?

@haraldh opinion on this?

**Comment by poettering** [comment]
I talked to @haraldh, he is fine with such a change. 

I filed the following bug against dracut in this context, since it lacks DefaultDependencies=no, which needs to be fixed before we can make this change:

https://github.com/dracutdevs/dracut/issues/151

**Comment by arvidjaar** [comment]
> This should solve the issue you raise and make a lot of things easier, no?

This still means that every unit with `DefaultDependencies=no` must have this Before and Conflicts added to it; and all units so far known to trigger this issue have `DefaultDependencies=no`. So no, sorry, I am afraid it won't fix this issue.

**Comment by abhishpaliwal** [comment]
Is there any update on this issue I am also seeing the same issue at my target.

error message "System is booting up. See pam_nologin(8)"

**Comment by timtadh** [comment]
I am also seeing this issue after I installed the systemd package on Ubuntu server 14.04 which locked me out of all of my servers. Pretty unexpected to run a script to apt-get install a package and then not be able to login on SSH to any of them.

**Comment by adepretis** [comment]
Same problem as @timtadh described: after installing the ``systemd`` package on ``Ubuntu Server 14.04.05``, logging in via ssh was no longer possible:

```
 $ ssh myserver
System is booting up.
```

Removing ``systemd`` and a reboot fixed the problem!

**Comment by jrmithdobbs** [comment]
Real solution: revert your systems to a SysV or other functional init system. Might I suggest runit? All the features and none of the bullshit, posturing, or ignored bug reports.
